### ISSUE TITLE
Enforcing things about class members

### ISSUE THREAD

**Issue by NickHarvey2** [BenMorris/NetArchTest#4]
I've been using NetArchTest in several projects and recently ran into an issue where some DTOs had non-nullable primitive types as fields. I wanted to be able to enforce that all our DTO classes fields and properties were nullable types.

Since this isn't something NetArchTest can currently do, I was considering submitting a PR that added this capability, as well as the ability to validate other qualities of members of classes (e.g. that they are readonly, or what their access level is, or other things).

Does this sound useful? And if so, do you have suggestions for other qualities of members that it would make sense to be able to validate? 

As far as syntax, I was thinking of trying for something like
`Members.OfTypes().InNamespace("Project.Models").That().AreFields().Or().AreProperties().Should().BeNullable();`
or
`Members.OfTypes(Types.InNamespace("Project.Models")).That().AreFields().Or().AreProperties().Should().BeNullable();`

**Comment by BenMorris** [comment]
This could indeed be useful, but it may get quite complex once we scratch the surface.

One thing to consider is whether or not we are straying into the territory of code review tools - is this really an *architectural* thing, or would it be better implemented via something like SonarQube?

I'm also struggling to come up with meaningful rules that you might want to apply accross members. It may be useful to check that a class is immutable (i.e. fields and properties are read-only). However, this could be achieved by adding an `IsImmutable()` function to the existing conditions and predicates.

Unless we can draw up a list of useful rules then your use case may be better answered via a function - e.g. `HasNullableMembers()`.

If we were to extend the syntax to allow for more flexible querying of members the second suggestion seems cleaner to me:

```
// Select some types
var types = Types.InCurrentDomain().That().ResideInNamespace("NetArchTest.SampleLibrary.Presentation");

// Implement a rule based on the members of these types
var result = Members.OfTypes(types).That().AreFields().Should().BeNullable().GetResult().IsSuccessful;
```

On the other hand, it may be easier to implement a "linking" method in both the type predicates and conditions, i.e.

```
var types = Types.InCurrentDomain().That().ResideInNamespace("Example").Should().HaveMembersThat().AreNullable();
```
A couple of extra considerations:

- If we are dealing with all the members then under the hood we'll need to traverse the collections of fields, properties, events, methods, etc. Perhaps it might be more meaningful to separate these types of member out? This would add further complexity to the implementation...
- Some fields are generated by the compiler as part of the async\await implementation. Would we want to ignore these?

**Comment by NickHarvey2** [comment]
Fair question regarding this being an architectural thing - in reality, it probably is not, but it does fit with the functionality of validating structure of code at test time.

I agree that it would be nice to check for something like this via an SCA tool like SonarQube (which I've also been trying to get adopted in my org, coincidentally), but I don't really know how to accomplish that - whether a given class is a DTO or not isn't always clear by looking at the class in isolation - a given POCO may be a DTO, or  it may be just an internal model used by the application, or something used by an ORM, etc.

Also agree that it is difficult to think of rules that you'd want to apply across members of any class more complex than a POCO. I will likely look into implementing `IsImmutable()` and `HasNullableMembers()`, which together should provide the functionality I'm looking for.

Regarding whether to ignore compiler generated fields, I honestly don't know enough about what exactly the compiler does under the hood to implement async/await functionality to answer, but will read up on it prior to working on any of this.

**Comment by BenMorris** [comment]
I have reviewed and released as v1.1.3 and pushed to NuGet. 

Thanks very much for the input here - these are some very useful new methods.

The only change I made was to rename one of the methods to OnlyHaveNullableMembers as it seemed to read better in the fluent API syntax.

**Comment by NickHarvey2** [comment]
great, thanks for the timely release! I'm already using this in a few projects
